syntax = "proto3";

package master;

import "common.proto";

service MasterService {
  rpc Healthcheck (HealthcheckRequest) returns (HealthcheckResponse) {}
  rpc GetWorkersToSaveBlob (GetWorkersToSaveBlobRequest) returns (GetWorkersToSaveBlobResponse) {}
  rpc NotifyBlobSaved (NotifyBlobSavedRequest) returns (NotifyBlobSavedResponse) {}
  rpc GetWorkerWithBlob (GetWorkerWithBlobRequest) returns (GetWorkerWithBlobResponse) {}
  rpc DeleteBlob (DeleteBlobRequest) returns (DeleteBlobResponse) {}
}

message HealthcheckRequest {}

message HealthcheckResponse {
  string message = 1;
}

// Message send by front end to get workers IPs and ports
message GetWorkersToSaveBlobRequest {
  string blob_hash = 1;
  uint64 size_mb = 2;
}

message GetWorkersToSaveBlobResponse {
  repeated string addresses = 1; // addresses can be IP:PORT, but also DNS_NAME:PORT
}

// Message send by worker to notify successful saving of blob
message NotifyBlobSavedRequest {
  string worker_id = 1;
  string blob_hash = 2;
}

message NotifyBlobSavedResponse {}

// Message send by frontend to get address of worker with specific blob
message GetWorkerWithBlobRequest {
  string blob_hash = 1;
}

message GetWorkerWithBlobResponse {
  string addresses = 1;
}

// Message send by frontend to request deletion of a blob
message DeleteBlobRequest {
  string blob_hash = 1;
}

message DeleteBlobResponse {}